rust/gitxetcore/src/command/dematerialize.rs

```diff
@@ -35,31 +36,75 @@ pub async fn dematerialize_command(cfg: XetConfig, args: &DematerializeArgs) ->
     let mut lazyconfig =
         LazyPathListConfigFile::load_smudge_list_from_file(&lazy_config_path, false).await?;
 
-    let path_list = list_files_from_repo(&repo.repo, &args.path, None, args.recursive)?;
+    let workdir_root = repo.repo_dir.clone();
+
+    // now they are relative path to the working directory root
+    let path_list = walk_working_dir(&workdir_root, &args.path, args.recursive)?;
+
+    let path_list = filter_files_from_index(&path_list, repo.repo.clone())?;
 
     if path_list.is_empty() {
         eprintln!(
-            "Didn't find any files under {}, skip dematerializing.",
+            "Didn't find any files under {:?}, skip dematerializing.",
             args.path
         );
         return Ok(());
     }
 
-    eprintln!("Dematerializing {} files...", path_list.len());
+    eprintln!("Dematerializing {} file(s)...", path_list.len());
 
-    let path_list_ref: Vec<_> = path_list.iter().map(|s| s.as_str()).collect();
+    let path_list_ref: Vec<_> = path_list
+        .iter()
+        .map(|s| s.to_str().unwrap_or_default())
+        .collect();
 
     lazyconfig.drop_rules(&path_list_ref);
 
     // saves changes to the file
     drop(lazyconfig);
 
-    // rerun smudge filter
-    std::fs::remove_file(cfg.repo_path()?.join("index"))?;
-
-    run_git_captured(None, "checkout", &["--force"], true, None)?;
+    // clean all files
+    let absolute_path_list: Vec<_> = path_list.iter().map(|p| workdir_root.join(p)).collect();
+
+    let translator = Arc::new(PointerFileTranslator::from_config(&cfg).await?);
+    let translator_ref = &translator;
+
+    tokio_par_for_each(
+        absolute_path_list,
+        MAX_CONCURRENT_UPLOADS,
+        |path, _| async move {
+            let translator = translator_ref.clone();
+            clean_file_to_itself(&translator, &path).await
+        },
+    )
+    .await
+    .map_err(|e| match e {
+        parutils::ParallelError::JoinError => {
+            anyhow::anyhow!("Join error on clean file")
+        }
+        parutils::ParallelError::TaskError(e) => e,
+    })?;
+
+    // update index so dematerialized files don't show as "Changes not staged for commit"
+    repo.run_git_checked_in_repo("add", &["-u"])?;
```

Comment:
Yes. I didn't have this initially, found out during hand test.

Reply:
Hmmm.  Our criteria for whether or not to smudge a file is that it's either pure binary or text under like 256K, then I wonder if it's hitting something around those limits.  I think it's an open question of what to do when dematerialize is called on a file that would normally be stored as the full file.   In this context, it would show up as changed then.